New task-scheduling infrastructure.

RonConescu opened the pull request:
Take 2.

This commit contains a major architectural revision:  cleaning up our scheduling system.  The original system scheduled tasks into the future; this system maintains the scheduling rules, but doesn't actually schedule anything until the user taps on it.

We also:
- centralized lots of code, making it easier to maintain and debug
- made certain common features (schedule delays and task-expiration periods) data-driven instead of software-driven

This commit addresses concerns raised in the comments to my previous commit:
- DataSubstrate has been re-consolidated (and extended)
- superfluous files have been removed
- formatting has (mostly) been preserved

Commit:
New task-scheduling infrastructure.

This commit contains a major architectural revision:  cleaning up our scheduling system.  The original system scheduled tasks into the future; this system maintains the scheduling rules, but doesn't actually schedule anything until the user taps on it.  We also:
- centralized lots of code, making it easier to maintain and debug
- made certain common features (schedule delays and task-expiration periods) data-driven instead of software-driven

jwe-apple commented:
Please remove OpenSSL.

jwe-apple commented:
@p2 Could you please have a look at this from your point of view?

p2 commented:
Hmm, it doesn't seem to be up to date with latest master. This commit re-introduces several changes made in the past few weeks, such as re-separation of `APCDataSubstrate` (#30), reverses fixes to storyboards and reverses a lot of fixed code style issues.

jwe-apple commented:
@RonConescu I assume you're going to update this before we review? I emailed Pablo earlier, did that get through?

peculiar commented:
@jwe-apple Your comments were received and this PR will be updated.

Commit:
New task-scheduling infrastructure, take 2.

This commit contains a major architectural revision:  cleaning up our scheduling system.  The original system scheduled tasks into the future; this system maintains the scheduling rules, but doesn't actually schedule anything until the user taps on it.

We also:
- centralized lots of code, making it easier to maintain and debug
- made certain common features (schedule delays and task-expiration periods) data-driven instead of software-driven

This commit addresses concerns raised in the pull-request comments to my previous commit:
- DataSubstrate has been re-consolidated (and extended)
- superfluous files have been removed
- formatting has (mostly) been preserved

RonConescu commented:
Hi, folks.  I've submitted a second commit.

Commit:
Code cleanup.

Commit:
Code cleanup.

Commit:
Adding documentation.

Commit:
Adding documentation.

RonConescu commented:
@jwe-apple , @YuanZhu-apple , @p2 -- I completely redid the contents of the pull request, and updated the introductory comment.  I've also submitted a bunch of subsequent edits to comments and formatting.  Since I haven't seen any comments from you, I'll assume that GitHub hasn't sent you an email about any of that... ?  I look forward to your next round of feedback.

YuanZhu-apple commented on APCAppCore/APCAppCore/DataMonitor/APCDataMonitor+Bridge.m:
```diff
@@ -39,32 +39,47 @@
+             {
+                 [self finishRefreshCommandPassingError: errorFromServerFetch
+                                      toCompletionBlock: completionBlock];
+             }
```
Declare a weakSelf outside this block and use it here to avoid reference loop. 
`__weak typeof(self) weakSelf = self;`

RonConescu replied:
Done, in commit 5c46280742268fc22db537017c1e28368589654d.

YuanZhu-apple commented on APCAppCore/APCAppCore/DataMonitor/APCDataMonitor+Bridge.m:
```diff
@@ -39,32 +39,47 @@
-        }];
+                     [self finishRefreshCommandPassingError: errorFromRefreshingSurveys
+                                          toCompletionBlock: completionBlock];
+
```
Same here, weakSelf

RonConescu replied:
Done, in commit 5c46280742268fc22db537017c1e28368589654d.

p2 commented on APCAppCore/APCAppCore/DataSubstrate/Model/APCSchedule+Bridge.h:
```diff
@@ -33,6 +33,3 @@
  
 #import "APCSchedule.h"
```
It seems this file is now empty, save for the license and import, can it be removed?

RonConescu replied:
Done, in commit 5c46280742268fc22db537017c1e28368589654d.

p2 commented on APCAppCore/APCAppCore/DataSubstrate/Model/APCUser.h:
```diff
@@ -118,4 +118,24 @@ typedef NS_ENUM(NSInteger, APCUserConsentSharingScope) {
+ 
+ These days, we track the date the user signs up.  In
+ earlier versions of the apps, we didn't.  This method
+ represents a set of nest-best-guesses about that date,
```
Typo _nest_ instead of _next_.

RonConescu replied:
Done, in commit 5c46280742268fc22db537017c1e28368589654d.

p2 commented on APCAppCore/APCAppCore/Library/Categories/SBBGuidCreatedOnVersionHolder+APCAdditions.h:
```diff
@@ -32,9 +32,3 @@
 // 
  
```
This and its implementation counterpart now are also empty, to be removed entirely?

RonConescu replied:
Done, in commit 5c46280742268fc22db537017c1e28368589654d.

p2 commented:
Sorry for the delay @RonConescu . I did not test the code, but from going over it in GitHub this looks all good from my side – I'm mostly focused on keeping interdependencies down, and there are improvements on that front in this PR.

(One hint: [coding-style-guide.md](https://github.com/rsanchezsaez/ResearchKit/blob/2febdc155a0fbee72a759dff82cd5af93733b923/docs-standalone/coding-style-guide.md))

YuanZhu-apple commented on APCAppCore/APCAppCore/Library/Categories/NSDate+Helper.h:
```diff
@@ -77,6 +94,138 @@ extern NSString * const NSDateDefaultDateFormat;
+ @see -dateByAddingISO8601Duration:
+ */
+- (NSDate *) dateBySubtractingISO8601Duration: (NSString *) durationString;
 
```
Generally, I think all these utility function need test coverage.

RonConescu replied:
Created a [GitHub issue](https://github.com/ResearchKit/AppCore/issues/75) for this.

RonConescu commented:
John, Pascal, Yuan,

Thank you for the comments above.  A general question:  for many of the pull requests on this project, you typically have many comments and requirements, both granular and architectural.  Pascal and Yuan's comments above seem more granular than architectural, which suggested to me that I should wait to make more changes until you'd had a chance to add architectural requirements.  (Obviously, Pascal's "interdependencies" comment is architectural, but it was also a "looks good" instead of a "please change such-and-such.")  Should I continue to wait?  If I implement Pascal and Yuan's stuff, above, will this pull request become acceptable?

Thanks, very much.

Ron

(to/cc:  @jwe-apple @p2 @YuanZhu-apple @insha @peculiar @pabloarce-yml)

Commit:
Responding to pull-request comments.

RonConescu commented:
Gang,

I've submitted a new commit, with responses to each item above.  I look forward to your feedback.

Ron

cc:  @jwe-apple @p2 @YuanZhu-apple @insha @peculiar @pabloarce-yml @RonConescu

p2 commented:
I'm not set up to test the new scheduler just yet, hence no substantial comments from my side and no objections either.

YuanZhu-apple commented on APCAppCore/APCAppCore/Library/Scheduler/APCScheduleIntervalEnumerator.m:
```diff
@@ -0,0 +1,266 @@
+    if (serializedTimesOfDayString.length > 0)
+    {
+        NSDateFormatter *formatter = [NSDateFormatter new];
+        formatter.locale = [NSLocale localeWithLocaleIdentifier: @"en_US_POSIX"];
```
Use kAPCDateFormatLocaleEN_US_POSIX

RonConescu replied:
Done. Also applied that constant to everywhere else in AppCore which needed it.

YuanZhu-apple commented:
@RonConescu I think this is ready to go after some response for the last comment I made.

Commit:
Responding to pull-request comments.

RonConescu commented:
@YuanZhu-apple — Back to you, sir.
